### PR TITLE
Improve URL validation disallow use of IP address

### PR DESCRIPTION
**Pull request by MKodde**
To get a bit more control over the validator, the regex was swapped in
favour of using the javascript url parser combined with an additional
ip validator.

Disclaimer: the build is marked 'not complete yet' even though Travis finished with success.. :man_shrugging: 

https://www.pivotaltracker.com/story/show/172833472

### CODE REVIEW

File: app/js/components/validator_helper.ts [hunk 1]
```diff
@@ -5,9 +5,20 @@ export class ValidatorHelper {
...
-
-    return regExp.test(value);
+    const ipValidator = require('ip-validator');
+    try {
+      // The url should be valid according to the JS parser
+      const url: URL = new URL(value);
```
Comment (pablothedude) [inline app/js/components/validator_helper.ts:11]: Is it safe to use URL or was it intended to drop IE11 support?
`https://caniuse.com/#search=URL`

File: app/js/components/validator_helper.ts [hunk 2]
```diff
@@ -5,9 +5,20 @@ export class ValidatorHelper {
...
+      // Url's can not be IP addresses, except for loopback addresses, they are verified in 'validateLoopback'
+      const isIp: boolean = ipValidator.ipv4(url.host) || ipValidator.ipv6(url.host);
+      // Localhost is not allowed in this validator. The validateLoopback will verify this as valid
+      const isLocalhost: boolean = url.host === 'localhost';
+      // Only allow http(s) protocol
+      const isInvalidProtocol: boolean = !['http:', 'https:'].includes(url.protocol);
```
[the remark this answers is not in the document]
Comment (MKodde) [reply app/js/components/validator_helper.ts]: :+1:  did this

File: app/js/components/validator_helper.ts [hunk 3]
```diff
@@ -5,9 +5,20 @@ export class ValidatorHelper {
...
-
-    return regExp.test(value);
+    const ipValidator = require('ip-validator');
```
[the remark this answers is not in the document]
Comment (MKodde) [reply app/js/components/validator_helper.ts]: Not sure my use of require was incorrect. you should be able to 'import' a function from node modules like this right? Other members of this class where not depending on that function. 

I ended up removing the library, and only using the regexes from that library to perform the logic in a private function.

File: app/js/components/validator_helper.ts [hunk 4]
```diff
@@ -5,9 +5,20 @@ export class ValidatorHelper {
...
+    try {
+      // The url should be valid according to the JS parser
+      const url: URL = new URL(value);
+      // Url's can not be IP addresses, except for loopback addresses, they are verified in 'validateLoopback'
+      const isIp: boolean = ipValidator.ipv4(url.host) || ipValidator.ipv6(url.host);
```
[the remark this answers is not in the document]
Comment (MKodde) [reply app/js/components/validator_helper.ts]: Yes, all the type definitions are somewhat excessive. I added them initially to satisfy my TS checker which might be configured somewhat strict.

File: app/js/components/validator_helper.ts (hunk 1 above)
Comment (MKodde) [reply app/js/components/validator_helper.ts]: Yeah, we can drop IE11 support

File: app/js/components/validator_helper.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply app/js/components/validator_helper.ts]: You don't have to do this in TS and you don't have to do this in PHP either, but I think it's a good practice to have all your imports at the top so you can easily see if an external dependency is being used.

I love your solution of just using the regex!

**Pull request merged**